execbuilder: consider disabling stats forecasts for execbuilder tests

pkg/sql/opt/exec/execbuilder/tests/local/local_test.TestExecBuild_explain [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8672059?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8672059?buildTab=artifacts#/) on release-22.2 @ [beb24ea0688dcb96ca8a1f2ee85cd154efe0c0b2](https://github.com/cockroachdb/cockroach/commits/beb24ea0688dcb96ca8a1f2ee85cd154efe0c0b2):


```
                │
                └── • group (hash)
                    │ estimated row count: 3
                    │ group by: username
                    │
                    └── • sort
                        │ estimated row count: 3
                        │ order: +"role"
                        │
                        └── • hash join (left outer)
                            │ estimated row count: 3
                            │ equality: (username) = (member)
                            │ left cols are key
                            │
                            ├── • group (hash)
                            │   │ estimated row count: 3
                            │   │ group by: username
                            │   │
                            │   └── • window
                            │       │ estimated row count: 3
                            │       │
                            │       └── • render
                            │           │
                            │           └── • hash join (left outer)
                            │               │ estimated row count: 3
                            │               │ equality: (username) = (username)
                            │               │ left cols are key
                            │               │
                            │               ├── • scan
                            │               │     estimated row count: 3 (100% of the table; stats collected <hidden> ago; using stats forecast)
                            │               │     table: users@users_user_id_idx
                            │               │     spans: FULL SCAN
                            │               │
                            │               └── • scan
                            │                     estimated row count: 1 (100% of the table; stats collected <hidden> ago)
                            │                     table: role_options@primary
                            │                     spans: FULL SCAN
                            │
                            └── • scan
                                  estimated row count: 1 (100% of the table; stats collected <hidden> ago)
                                  table: role_members@role_members_role_idx
                                  spans: FULL SCAN
[11:48:33] --- progress: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3030/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/explain: 50 statements
[11:48:33] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3030/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/explain with config local: 50 tests, 1 failures
    logic.go:3785: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3030/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/explain:697: error while processing
    logic.go:3785: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3030/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/sql/opt/exec/execbuilder/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/opt/exec/execbuilder/testdata/explain:697: too many errors encountered, skipping the rest of the input
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/0f5ae97386d181c4ea2f7c775e9c3a67/logTestExecBuild_explain1282353941
--- FAIL: TestExecBuild_explain (46.10s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExecBuild_explain.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-24453

--- a/pkg/sql/logictest/logic.go
+++ b/pkg/sql/logictest/logic.go
@@ -1677,6 +1677,14 @@ func (t *logicTest) newCluster(
 			t.Fatal(err)
 		}
 
+		// We also disable stats forecasts to have deterministic tests. See #97003
+		// for details.
+		if _, err := conn.Exec(
+			"SET CLUSTER SETTING sql.stats.forecasts.enabled = false",
+		); err != nil {
+			t.Fatal(err)
+		}
+
 		// Update the default AS OF time for querying the system.table_statistics
 		// table to create the crdb_internal.table_row_statistics table.
 		if _, err := conn.Exec(

--- a/pkg/sql/opt/exec/execbuilder/testdata/forecast
+++ b/pkg/sql/opt/exec/execbuilder/testdata/forecast
@@ -5,6 +5,13 @@
 # Verify that we create and use statistics forecasts for a simple table that
 # grows at a constant rate.
 
+let $forecastsEnabledPrev
+SHOW CLUSTER SETTING sql.stats.forecasts.enabled
+
+# First: enable statistics forecasts.
+statement ok
+SET CLUSTER SETTING sql.stats.forecasts.enabled = true
+
 statement ok
 CREATE TABLE g (a INT PRIMARY KEY) WITH (sql_stats_automatic_collection_enabled = false)
 
@@ -1596,3 +1603,7 @@ RESET enable_zigzag_join
 
 statement ok
 RESET optimizer_use_forecasts
+
+# Finally, restore forecasts setting to its previous value.
+statement ok
+SET CLUSTER SETTING sql.stats.forecasts.enabled = $forecastsEnabledPrev

--- a/pkg/sql/opt/exec/execbuilder/testdata/forecast1401
+++ b/pkg/sql/opt/exec/execbuilder/testdata/forecast1401
@@ -2,6 +2,13 @@
 
 # Test whether statistics forecasts help the query from case 1401.
 
+let $forecastsEnabledPrev
+SHOW CLUSTER SETTING sql.stats.forecasts.enabled
+
+# First: enable statistics forecasts.
+statement ok
+SET CLUSTER SETTING sql.stats.forecasts.enabled = true
+
 statement ok
 CREATE TABLE t1401 (
   a INT8 NOT NULL,
@@ -18481,3 +18488,7 @@ vectorized: true
               estimated row count: 101 - 632 (<0.01% of the table; stats collected <hidden> ago; using stats forecast)
               table: t1401@t1401_f_c_idx1 (partial index)
               spans: [/4458256 - /4458256/'2022-01-25 21:17:27.216095+00']
+
+# Finally, restore forecasts setting to its previous value.
+statement ok
+SET CLUSTER SETTING sql.stats.forecasts.enabled = $forecastsEnabledPrev

--- a/pkg/sql/opt/exec/execbuilder/testdata/partial_stats
+++ b/pkg/sql/opt/exec/execbuilder/testdata/partial_stats
@@ -9,6 +9,13 @@
 # We store multiple full stats to ensure that merged stats
 # work with forecasted statistics as well.
 
+let $forecastsEnabledPrev
+SHOW CLUSTER SETTING sql.stats.forecasts.enabled
+
+# First: enable statistics forecasts.
+statement ok
+SET CLUSTER SETTING sql.stats.forecasts.enabled = true
+
 statement ok
 CREATE TABLE g (b INT PRIMARY KEY) WITH (sql_stats_automatic_collection_enabled = false)
 
@@ -1250,3 +1257,7 @@ statistics_name  column_names  created                                 row_count
 full_b           {b}           2022-12-12 15:22:21.988179 +0000 +0000  28         26              3           3
 partial_b        {b}           2022-12-13 15:22:21.988179 +0000 +0000  36         32              5           3
 __merged__       {b}           2022-12-13 15:22:21.988179 +0000 +0000  61         57              5           3
+
+# Finally, restore forecasts setting to its previous value.
+statement ok
+SET CLUSTER SETTING sql.stats.forecasts.enabled = $forecastsEnabledPrev

